File: canVisitAllRooms.java
import java.util.*;
// Approach 1- Iterative DFS [using explicit stack]
class Solution{ // TC O(N+E) SC O(N)
	public boolean canVisitAllRooms(List<List<Integer>> rooms){
		boolean[] seen= new boolean[rooms.size()];
		Stack<Integer> stack= new Stack<>();
		seen[0]=true;
		stack.push(0);
		while(!stack.isEmpty()){
			int room=stack.pop();
			for(int key: rooms.get(room)){
				if(!seen[key]){
					seen[key]=true;
					stack.push(key);
				}
			}
		}

		for(boolean visited:seen){
			if(!visited) return false;
		}
		return true;
	}
}

class canVisitAllRooms{
	public static void main(String[] args) {
		List<List<Integer>> rooms= new ArrayList<>(Arrays.asList(Arrays.asList(1),
			Arrays.asList(2),Arrays.asList(3),Arrays.asList()));
		Solution s= new Solution();
		System.out.println(s.canVisitAllRooms(rooms));
	}
}

File: canVisitAllRooms.py
# Approach 1- Iterative DFS [using explicit stack]
def canVisitAllRooms(rooms): # TC O(N+E) // SC O(N)
	seen=[False]*len(rooms)
	stack=[]
	seen[0]=True 
	stack.append(0)
	while stack:
		room=stack.pop()
		for key in rooms[room]:
			if not seen[key]:
				seen[key]=True
				stack.append(key)

	for visited in seen:
		if not visited:
			return False 
	return True 

	# return all(seen) Return true iff we've visited every room

print(canVisitAllRooms([[1],[2],[3],[]]))
